src/helics/application_api/Inputs.hpp

```diff
@@ -368,60 +367,133 @@ class InputT : public InputBase
+        value_callback (out, time);
+    }
+};
+
+template <class X>
+void Input::getValue_impl (std::integral_constant<int, 0> /*V*/, X &out)
+{
+    if (fed->isUpdated (*this) || (hasUpdate && !changeDetectionEnabled))
+    {
+        auto dv = fed->getValueRaw (*this);
+        if (type == helics_type_t::helicsUnknown)
         {
-            changeDetectionEnabled = true;
+            type = getTypeFromString (fed->getPublicationType (*this));
         }
-        delta = deltaV;
-        if (delta < 0.0)
+
+        valueExtract (dv, type, out);
+        if (changeDetectionEnabled)
         {
-            changeDetectionEnabled = false;
+            if (changeDetected (lastValue, out, delta))
+            {
+                lastValue = make_valid (out);
+            }
+            else
+            {
+                valueExtract (lastValue, out);
+            }
+        }
+        else
+        {
+            lastValue = make_valid (out);
         }
     }
-    void enableChangeDetection (bool enabled = true) { changeDetectionEnabled = enabled; }
+    else
+    {
+        valueExtract (lastValue, out);
+    }
+    hasUpdate = false;
+}
 
-    /** store the value in the given variable
-    @param[out] out the location to store the value
-    */
-    const X &operator[] (int index) const { return fed->getValue<X> (id, index); }
-  private:
-    void handleCallback (Time time)
+template <class X>
+inline const X &getValueRefImpl (defV &val)
+{
+    valueConvert (val, helicsType<X> ());
+    return mpark::get<X> (val);
+}
+
+template <>
+inline const std::string &getValueRefImpl (defV &val)
+{
+    // don't convert a named point to a string
+    if ((val.index () == namedPointLoc))
     {
-        X out;
-        fed->getValue (id, out);
-        value_callback (out, time);
+        return mpark::get<named_point> (val).name;
     }
-};
+    else
+    {
+        valueConvert (val, helics_type_t::helicsString);
+        return mpark::get<std::string> (val);
+    }
+}
+
+template <class X>
+const X &Input::getValueRef ()
+{
+    static_assert (std::is_same<typeCategory<X>, std::integral_constant<int, 0>>::value,
+                   "calling getValue By ref must be with a primary type");
+    if (fed->isUpdated (*this) || (hasUpdate && !changeDetectionEnabled))
+    {
+        auto dv = fed->getValueRaw (*this);
+        if (type == helics_type_t::helicsUnknown)
+        {
+            type = getTypeFromString (fed->getPublicationType (*this));
+        }
+
+        if (changeDetectionEnabled)
+        {
+            X out;
+            valueExtract (dv, type, out);
+            if (changeDetected (lastValue, out, delta))
+            {
+                lastValue = make_valid (std::move(out));
+            }
+        }
+        else
+        {
+            valueExtract (dv, type, lastValue);
+        }
+    }
+    else
+    {
+		//TODO:: PT make some logic that it can get the raw data from the core again if it was converted already
```

Comment:
No, it was actually an issue before, I just made a note of it now and is a little obvious with the getValueRef function., but it would be pretty odd use case, and it is one of the things I hope to clear up after the beta comes out.